components/shared/net/request.rs

```diff
@@ -598,6 +599,111 @@ impl Request {
             ResourceTimingType::Resource
         }
     }
+
+    /// <https://fetch.spec.whatwg.org/#concept-request-tainted-origin>
+    fn has_redirect_tainted_origin(&self) -> bool {
+        // https://github.com/whatwg/fetch/issues/1773
+        let Origin::Origin(request_origin) = &self.origin else {
+            panic!("origin cannot be \"client\" at this point in time");
+        };
+
+        // Step 1. Let lastURL be null.
+        let mut last_url = None;
+
+        // Step 2. For each url of request’s URL list:
+        for url in &self.url_list {
+            // Step 2.1 If lastURL is null, then set lastURL to url and continue.
+            let Some(last_url) = &mut last_url else {
+                last_url = Some(url);
+                continue;
+            };
+
+            // Step 2.2 If url’s origin is not same origin with lastURL’s origin and
+            //          request’s origin is not same origin with lastURL’s origin, then return true.
+            if url.origin() != last_url.origin() && *request_origin != last_url.origin() {
+                return true;
+            }
+
+            // Step 2.3 Set lastURL to url.
+            *last_url = url;
+        }
+
+        // Step 3. Return false.
+        false
+    }
+
+    /// <https://fetch.spec.whatwg.org/#serializing-a-request-origin>
+    fn serialize_request_origin(&self) -> headers::Origin {
+        // 1. If request has a redirect-tainted origin, then return "null".
+        if self.has_redirect_tainted_origin() {
+            return headers::Origin::NULL;
+        }
+
+        // 2. Return request’s origin, serialized.
+        // NOTE: https://github.com/whatwg/fetch/issues/1773
+        let Origin::Origin(origin) = &self.origin else {
+            panic!("origin cannot be \"client\" at this point in time");
+        };
+
+        immutable_origin_to_hyper_origin(origin)
+    }
+
+    /// <https://fetch.spec.whatwg.org/#append-a-request-origin-header>
+    pub fn append_a_request_origin_header(&mut self) {
```

Comment:
I don't think this belongs here, because it is only used in `components/net`, or do you think other components will need it in the future? If not, this can be a function that takes a request as argument.